File: examples/cpp/quickstart.cpp
/**
 * @file quickstart.cpp
 * @brief Minimal working example of AuraIO C++ async read
 *
 * Build: make -C examples cpp-examples
 * Run:   ./examples/cpp/quickstart
 */

#include <auraio.hpp>

#include <iostream>
#include <fstream>
#include <cstring>
#include <fcntl.h>
#include <unistd.h>

constexpr size_t BUF_SIZE = 4096;

int main() {
    const char* test_file = "/tmp/auraio_quickstart_cpp.tmp";
    const char* test_data = "Hello from AuraIO C++! This is async I/O.\n";

    try {
        // Create a test file with known content
        {
            std::ofstream out(test_file, std::ios::binary);
            if (!out) {
                std::cerr << "Failed to create test file\n";
                return 1;
            }
            out.write(test_data, static_cast<std::streamsize>(strlen(test_data)));
        }

        // Create AuraIO engine (RAII - automatically cleaned up)
        auraio::Engine engine;

        // Allocate aligned buffer (RAII - automatically returned to pool)
        auto buffer = engine.allocate_buffer(BUF_SIZE);
        std::memset(buffer.data(), 0, BUF_SIZE);

        // Open file for reading
        int fd = open(test_file, O_RDONLY);
        if (fd < 0) {
            throw auraio::Error(errno, "open");
        }

        // Track completion with lambda capture
        bool done = false;
        ssize_t bytes_read = 0;

        // Submit async read with lambda callback
        engine.read(fd, buffer, BUF_SIZE, 0, [&](auraio::Request&, ssize_t result) {
            std::cout << "Read completed: " << result << " bytes\n";
            bytes_read = result;
            done = true;
        });

        // Wait for completion
        while (!done) {
            engine.wait(100);
        }

        // Verify result
        if (bytes_read > 0) {
            std::cout << "Data read: " << static_cast<char*>(buffer.data());
        }

        // Cleanup fd (buffer and engine cleaned up automatically)
        close(fd);
        unlink(test_file);

        std::cout << "Success!\n";
        return 0;

    } catch (const auraio::Error& e) {
        std::cerr << "AuraIO error: " << e.what() << "\n";
        unlink(test_file);
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        unlink(test_file);
        return 1;
    }
}

File: examples/rust/examples/quickstart.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 AuraIO Contributors


//! Minimal working example of AuraIO async read
//!
//! This example demonstrates the basic usage pattern:
//! 1. Create an engine
//! 2. Allocate a buffer
//! 3. Submit an async read with a callback
//! 4. Wait for completion
//!
//! Run: cargo run --example quickstart

use aura::{Engine, Result};
use std::fs::{self, File};
use std::io::Write;
use std::os::unix::io::AsRawFd;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

const BUF_SIZE: usize = 4096;

fn main() -> Result<()> {
    let test_file = "/tmp/aura_rust_quickstart.tmp";
    let test_data = b"Hello from AuraIO Rust bindings! This is async I/O.\n";

    // Create a test file with known content
    {
        let mut file = File::create(test_file).expect("Failed to create test file");
        file.write_all(test_data)
            .expect("Failed to write test data");
    }

    // Create AuraIO engine
    let engine = Engine::new()?;

    // Allocate aligned buffer from engine's pool
    let mut buf = engine.allocate_buffer(BUF_SIZE)?;

    // Zero the buffer
    buf.as_mut_slice().fill(0);

    // Open file for reading
    let file = File::open(test_file).expect("Failed to open test file");
    let fd = file.as_raw_fd();

    // Track completion state
    let done = Arc::new(AtomicBool::new(false));
    let done_clone = done.clone();

    // Capture result in callback
    let result_bytes = Arc::new(std::sync::atomic::AtomicIsize::new(0));
    let result_clone = result_bytes.clone();

    // Submit async read
    println!("Submitting async read...");
    unsafe {
        engine.read(fd, (&buf).into(), BUF_SIZE, 0, move |result| {
            match result {
                Ok(n) => {
                    println!("Read completed: {} bytes", n);
                    result_clone.store(n as isize, Ordering::SeqCst);
                }
                Err(e) => {
                    eprintln!("Read failed: {}", e);
                    result_clone.store(-1, Ordering::SeqCst);
                }
            }
            done_clone.store(true, Ordering::SeqCst);
        })?;
    }

    // Wait for completion
    while !done.load(Ordering::SeqCst) {
        engine.wait(100)?;
    }

    // Verify result
    let bytes_read = result_bytes.load(Ordering::SeqCst);
    if bytes_read > 0 {
        let content = std::str::from_utf8(&buf.as_slice()[..bytes_read as usize])
            .unwrap_or("<invalid utf-8>");
        println!("Data read: {}", content);
    }

    // Cleanup test file
    fs::remove_file(test_file).ok();

    println!("Success!");
    Ok(())
}
